app/src/main/java/com/kpiroom/bubble/ui/accountSetup/AccountSetupLogic.kt

```diff
@@ -0,0 +1,195 @@
+        clearUsernameFocus.value = true
+        progress.apply {
+            username.value.let {
+
+                AsyncProcessor {
+                    progress.loadAsync()
+                    when {
+                        it.isNullOrBlank() ->
+                            alertAsync(str(R.string.setup_username_cannot_be_empty))
+
+                        it.length < 6 ->
+                            alertAsync(str(R.string.setup_username_too_short))
+
+                        it.length > 16 ->
+                            alertAsync(str(R.string.setup_username_too_long))
+
+                        !it.matches(usernameRegex) ->
+                            alertAsync(str(R.string.setup_username_invalid))
+
+                        Source.api.usernameExists(it) ->
+                            alertAsync(str(R.string.setup_username_exists))
+
+                        !(isPhotoSet && isWallpaperSet) ->
+                            progress.alertAsync(
+                                warningPhotosUnchanged,
+                                ::isSetupFinishRequested
+                            )
+
+                        else -> {
+                            finishRequested.postValue(true)
+                            progress.finishAsync()
+                        }
+                    }
+                }
+            } handleError {
+                progress.alert(it.message)
+            } runWith (bag)
+        }
+    }
+
+    private val warningPhotosUnchanged
+        get() = app.getString(
+            R.string.setup_continue_with_unchanged,
+            when {
+                !(isPhotoSet || isWallpaperSet) -> str(R.string.setup_unchanged_photo_and_wallpaper)
+                !isPhotoSet -> str(R.string.setup_unchanged_photo)
+                !isWallpaperSet -> str(R.string.setup_unchanged_wallpaper)
+                else -> null
+            }
+        )
+
+    private fun isSetupFinishRequested(isRequested: Boolean) {
+        if (isRequested)
+            finishRequested.postValue(true)
+    }
+
+    fun finishSetup() {
+        progress.apply {
+            AsyncProcessor(Dispatchers.IO) {
+                loadAsync()
+
+                saveUserData()
+                uploadProfile()
+                setupComplete.postValue(true)
+
+                finishAsync()
+            } handleError {
+                alert(it.message)
+            } runWith (bag)
+        }
+    }
+
+    private suspend fun uploadProfile() = Source.apply {
+        userPrefs.apply {
+            uuid.let { id ->
+                api.uploadUserData(
+                    id,
+                    User(
+                        username,
+                        joinedDate,
+                        isPhotoSet,
+                        isWallpaperSet
+                    )
+                )
+            }
+        }
+        uploadUserImages()
+    }
+
+    private suspend fun uploadUserImages() = Source.apply {
+        val photoUri = photoUri.value
```

Comment:
in order to "clean" `photoUri` do the following:
`val photoUri = photoUri.value ?: return@apply`
The same thing for `wallpaperUri`